tool/formula/src/main/java/org/openscience/cdk/formula/IsotopePatternGenerator.java

```diff
@@ -98,122 +131,132 @@ public IsotopePattern getIsotopes(IMolecularFormula molFor) {
+    private IsotopePattern calculateAbundanceAndMass(IsotopePattern current, List<IsotopeContainer> additional) {
 
-        if (isotopes == null) return isotopePattern;
+        if (additional == null || additional.size() == 0) return current;
 
-        if (isotopes.length == 0) return isotopePattern;
-
-        double mass, previousMass, abundance, totalAbundance, newAbundance;
-
-        HashMap<Double, Double> isotopeMassAndAbundance = new HashMap<Double, Double>();
-        IsotopePattern currentISOPattern = new IsotopePattern();
-
-        // Generate isotopes for the current atom (element)
-        for (int i = 0; i < isotopes.length; i++) {
-            mass = isotopes[i].getExactMass();
-            abundance = isotopes[i].getNaturalAbundance();
-            currentISOPattern.addIsotope(new IsotopeContainer(mass, abundance));
-        }
+        List<IsotopeContainer>  containers = new ArrayList<>();
 
         // Verify if there is a previous calculation. If it exists, add the new
         // isotopes
-        if (isotopePattern == null) {
-            isotopePattern = currentISOPattern;
+        if (current == null) {
+            current = new IsotopePattern();
+            for (IsotopeContainer container : additional)
+                current.addIsotope(container);
         } else {
-            for (int i = 0; i < isotopePattern.getNumberOfIsotopes(); i++) {
-                totalAbundance = isotopePattern.getIsotopes().get(i).getIntensity();
-
-                if (totalAbundance == 0) continue;
-
-                for (int j = 0; j < currentISOPattern.getNumberOfIsotopes(); j++) {
-
-                    abundance = currentISOPattern.getIsotopes().get(j).getIntensity();
-                    mass = isotopePattern.getIsotopes().get(i).getMass();
-
-                    if (abundance == 0) continue;
-
-                    newAbundance = totalAbundance * abundance * 0.01;
-                    mass += currentISOPattern.getIsotopes().get(j).getMass();
-
-                    // Filter duplicated masses
-                    previousMass = searchMass(isotopeMassAndAbundance.keySet(), mass);
-                    if (isotopeMassAndAbundance.containsKey(previousMass)) {
-                        newAbundance += isotopeMassAndAbundance.get(previousMass);
-                        mass = previousMass;
+            for (IsotopeContainer container : current.getIsotopes()) {
+                for (IsotopeContainer other : additional) {
+
+                    double abundance = container.getIntensity() * other.getIntensity() * 0.01;
+                    double mass      = container.getMass() + other.getMass();
+
+                    // merge duplicates with some resolution
+                    IsotopeContainer existing = findExisting(containers, mass, resolution);
+                    if (existing != null) {
+                        existing.setMass((existing.getMass() + mass) / 2); // moving avg.
```

Comment:
I think it makes more sense to calculate a weighed average here (weighted by the isotope intensity). 
In other words, if I have 2 isotopes in the same container and isotope 1 has 100% intensity and isotope 2 has only 1% intensity, it makes sense to report the mass value closer to isotope 1 mass.